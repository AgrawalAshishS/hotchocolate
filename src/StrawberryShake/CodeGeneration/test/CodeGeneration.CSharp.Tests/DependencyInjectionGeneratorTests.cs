using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using Snapshooter.Xunit;
using Xunit;

namespace StrawberryShake.CodeGeneration.CSharp
{
    public class DependencyInjectionGeneratorTests
    {
        [Fact]
        public async Task GenerateModel()
        {
            // arrange
            var sb = new StringBuilder();
            var writer = new CodeWriter(sb);

            var generator = new DependencyInjectionGenerator();

            var descriptor = new DependencyInjectionDescriptor(
                "StarWarsClientServiceCollectionExtensions",
                "Demo",
                "StarWarsClient",
                "global::Demo.StarWarsClient",
                "global::Demo.IStarWarsClient",
                true,
                new List<string> { "Abc" },
                new List<string> { "Def" });

            // act
            await generator.WriteAsync(writer, descriptor);

            // assert
            sb.ToString().MatchSnapshot();
        }

        [Fact]
        public async Task GenerateModel_Without_Subscription()
        {
            // arrange
            var sb = new StringBuilder();
            var writer = new CodeWriter(sb);

            var generator = new DependencyInjectionGenerator();

            var descriptor = new DependencyInjectionDescriptor(
                "StarWarsClientServiceCollectionExtensions",
                "Demo",
                "StarWarsClient",
                "global::Demo.StarWarsClient",
                "global::Demo.IStarWarsClient",
                false,
                new List<string> { "Abc" },
                new List<string> { "Def" });

            // act
            await generator.WriteAsync(writer, descriptor);

            // assert
            sb.ToString().MatchSnapshot();
        }

        [Fact]
        public void CanHandle()
        {
            // arrange
            var generator = new DependencyInjectionGenerator();

            var descriptor = new DependencyInjectionDescriptor(
                "StarWarsClientServiceCollectionExtensions",
                "Demo",
                "StarWarsClient",
                "global::Demo.StarWarsClient",
                "global::Demo.IStarWarsClient",
                true,
                new List<string> { "Abc" },
                new List<string> { "Def" });

            // act
            var canHandle = generator.CanHandle(descriptor);

            // assert
            Assert.True(canHandle);
        }
    }
}
