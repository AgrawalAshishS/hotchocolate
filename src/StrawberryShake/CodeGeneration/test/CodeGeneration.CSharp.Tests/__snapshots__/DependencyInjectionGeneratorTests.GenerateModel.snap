using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection.Extensions;

namespace Demo
{
    [global::System.CodeDom.Compiler.GeneratedCode("StrawberryShake", "11.0.0")]
    public partial StarWarsClientServiceCollectionExtensions
    {
        private const string _clientName = "StarWarsClient";

        public static global::StrawberryShake.Configuration.IOperationClientBuilder AddStarWarsClient(this global::Microsoft.Extensions.DependencyInjection.IServiceCollection serviceCollection)
        {
            if (serviceCollection is null)
            {
                throw new global::System.ArgumentNullException(nameof(serviceCollection));
            }

            serviceCollection.AddSingleton<global::Demo.StarWarsClient, global::Demo.IStarWarsClient>();

            serviceCollection.AddSingleton<global::StrawberryShake.IOperationExecutorFactory>(sp =>
                new global::StrawberryShake.Http.HttpOperationExecutorFactory(
                    _clientName,
                    sp.GetRequiredService<global::System.Net.Http.IHttpClientFactory>().CreateClient,
                    sp.GetRequiredService<global::StrawberryShake.Configuration.IClientOptions>().GetOperationPipeline<global::StrawberryShake.Http.IHttpOperationContext>(_clientName),
                    sp.GetRequiredService<global::StrawberryShake.Configuration.IClientOptions>().GetOperationFormatter(_clientName),
                    sp.GetRequiredService<global::StrawberryShake.Configuration.IClientOptions>().GetResultParsers(_clientName)));

            serviceCollection.AddSingleton<global::StrawberryShake.IOperationStreamExecutorFactory>(sp =>
                new global::StrawberryShake.Http.Subscriptions.SocketOperationStreamExecutorFactory(
                    _clientName,
                    sp.GetRequiredService<global::StrawberryShake.Transport.ISocketConnectionPool>().RentAsync,
                    sp.GetRequiredService<global::StrawberryShake.Http.Subscriptions.ISubscriptionManager>(),
                    sp.GetRequiredService<global::StrawberryShake.Configuration.IClientOptions>().GetOperationFormatter(_clientName),
                    sp.GetRequiredService<global::StrawberryShake.Configuration.IClientOptions>().GetResultParsers(_clientName)));

            global::StrawberryShake.Configuration.IOperationClientBuilder builder = serviceCollection.AddOperationClientOptions(_clientName)
                .AddValueSerializer(() => new Abc())
                .AddResultParser(serializers => new Def(serializers))
            .AddOperationFormatter(serializers => new global::StrawberryShake.Http.JsonOperationFormatter(serializers))
            .AddHttpOperationPipeline(builder => builder.UseHttpDefaultPipeline());
            serviceCollection.TryAddSingleton<global::StrawberryShake.Http.Subscriptions.ISubscriptionManager, global::StrawberryShake.Http.Subscriptions.SubscriptionManager>();
            serviceCollection.TryAddSingleton<global::StrawberryShake.IOperationExecutorPool, global::StrawberryShake.OperationExecutorPool>();
            serviceCollection.TryAddEnumerable(new global::Microsoft.Extensions.DependencyInjection.ServiceDescriptor(
                typeof(global::StrawberryShake.Transport.ISocketConnectionInterceptor),
                typeof(global::StrawberryShake.Http.Subscriptions.MessagePipelineHandler),
                global::Microsoft.Extensions.DependencyInjection.ServiceLifetime.Singleton));

            return builder;
        }
    }
}
